osf/migrations/0197_draft_node_permissions.py

```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.15 on 2019-07-14 19:04
+from __future__ import unicode_literals
+
+import logging
+from django.db import migrations
+from django.core.management.sql import emit_post_migrate_signal
+
+logger = logging.getLogger(__name__)
+
+
+def post_migrate_signal(state, schema):
+    # this is to make sure that the draft registration permissions created earlier exist!
+    emit_post_migrate_signal(3, False, 'default')
```

Comment:
It looks like all calls to `update_permission_groups` are through the post_migrate signal in migrations. Making a direct call to `update_permission_groups` looks like it makes sense to me but I hesitate if there's something I'm missing about the reason why other calls to update the permissions groups are through the signal. 

What are your thoughts on why other calls for `update_permission_groups` are through the signal and why it would be fine for this case not to be, @aaxelb?

Reply:
well, there could be any number of `post_migrate` handlers that we may or may not be depending on. that's a big part of my objection to using signals this way -- *there's no way to know* without some digging.

...but here we are, so let's do some digging:
```
In [4]: from django.db import models

In [5]: models.signals.post_migrate
Out[5]: <django.dispatch.dispatcher.Signal at 0x7f7b94b6fc90>

In [6]: models.signals.post_migrate.receivers
Out[6]:
[(('django.contrib.auth.management.create_permissions', 140168763452928),
  <weakref at 0x7f7b9037a628; to 'function' at 0x7f7b91c92410 (create_permissions)>),
 ((140168703584688, 140168763452928),
  <weakref at 0x7f7b8f380db8; to 'function' at 0x7f7b91c929b0 (create_contenttypes)>),
 ((u'osf.apps.update_permissions_groups', 140168763452928),
  <weakref at 0x7f7b8f2b6b50; to 'function' at 0x7f7b91a801b8 (update_permission_groups)>)]
```
so we have three `post_migrate` receivers:
- `auth`'s [`create_permissions`](https://github.com/django/django/blob/1.11.15/django/contrib/auth/management/__init__.py#L40) -- add to `auth_permission` table for new models/permissions
- `contenttypes`'s [`create_contenttypes`](https://github.com/django/django/blob/1.11.15/django/contrib/contenttypes/management/__init__.py#L105) (pretty sure, anyway) -- add `ContentType`s for new models
- our own `update_permissions_groups` -- adds admin groups/permissions, and a set of groups/permissions for each provider

i can see why each of those might be required before mucking about with permissions in a migration (actually, we probably *don't* need `update_permissions_groups`), and emitting a signal is indeed the simplest solution.

i think it'd be cleaner, though, to make a util/helper like `ensure_permissions_and_contenttypes` that does exactly what we need -- it'd be easier to read/understand, and would avoid running irrelevant `post_migrate` handlers that might be added in the future.

because this is already a thing in a bunch of other migrations, though, i'm fine either way. your call! 👐

Reply:
That's fair - I'm inclined to stick with the signal for this PR for the sake of consistency and to avoid a potential rabbit hole. That being said, we can, and perhaps should, create a separate ticket to look into this more and create the utils for future migrations.